---
title: "GWAS on the A. thialana data set"
author: "Andrey Ziyatdinov"
date: "`r Sys.Date()`"
output:
  rmarkdown::html_vignette:
    fig_caption: yes
vignette: >
  %\VignetteIndexEntry{GWAS on the A. thialana data set}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

## Settings

### Parameters

```{r par}
chr <- 4

cores <- 2
```

### Packages

```{r inc}
library(dplyr)
library(Matrix)

library(athaliana)

library(lme4qtl)

library(solarius)
library(qqman)
```

## Load the data

### Phenotypes

```{r data_phen, cache = T}
phen <- athaliana_phen(traits = "FRI", rows_order = "snp")
```

```{r tab_phen, echo = FALSE, results = "asis"}
kable(head(phen))
```

```{r transofrm_phen}
phen <- mutate(phen, 
  inormal_FRI = solarius::transformTrait(FRI, "inormal"))
```

### Pre-computed GRM

```{r data_relmat, cache = T}
relmat <- athaliana_relmat()
```

```{r plot_relmat, fig.show = "hold", cache = TRUE, dependson = "data_relmat"}
image(Matrix(relmat))
image(Matrix(relmat[1:5, 1:5]))
```

```{r tab_relmat, echo = FALSE, results = "asis"}
kable(relmat[1:5, 1:5], digits = 2)
```

### Genotypes

```{r gdata, cache = TRUE}
gdat <- athaliana_snp(chr = chr)
  
gdat_subset <- gdat[1:20]
```

```{r tab_gdat, echo = FALSE, results = "asis"}
kable(head(select(gdat, 1:5)))
```

### Annotation

```{r annot, cache = T}
annot <- athaliana_annot()
```

## Polygenic model

```{r poly, cache = T}
(poly <- relmatLmer(inormal_FRI ~ (1|id), phen, relmat = list(id = relmat)))
```

## Association study of a small subset of snps

```{r assoc, cache = T}
#assoc <- assocLmer(FRI ~ (1|id), phen, relmat = list(id = relmat), data_snpcov = gdat_subset, batch_size = 10, cores = cores)
```


## Association study on all snps

```{r load_gwas, cache = T}
dir <- "/home/andrey/git/variani/athaliana/results/gwas/FRI"
load(file.path(dir, "gwas_rrblup_emmax.RData"))
load(file.path(dir, "gwas_rrblup_mm.RData"))
load(file.path(dir, "gwas_lmer_wald.RData"))
load(file.path(dir, "gwas_lmer_lr.RData"))
```

```{r join_gwas, cache = T}
tab <- data_frame(marker = gwas_rrblup_emmax$marker, 
  chrom = gwas_rrblup_emmax$chrom, pos = gwas_rrblup_emmax$pos,
  pval_emmax = 10^(-gwas_rrblup_emmax$log_FRI),
  pval_mm = 10^(-gwas_rrblup_mm$log_FRI))

tab <- bind_cols(tab, data_frame(snp = gwas_lmer_wald$snp, pval_wald = gwas_lmer_wald$pval))
tab <- bind_cols(tab, data_frame(pval_lr = gwas_lmer_lr$pval))
```  
  
```{r pval_diff}
tab <- mutate(tab,
  diff_logp_mm_wald = -log10(pval_mm) / -log10(pval_wald))
```

```{r show_diff}
tab %>% filter(diff_logp_mm_wald < 0.1)
```  

## Temp

```{r gwas, eval = F}
assoc <- assocLmer(FRI ~ (1|id), phen, relmat = list(id = relmat), data_snpcov = gdat, batch_size = 100, cores = cores)
gwas <- assocLmer(inormal_FRI ~ (1|id), phen, relmat = list(id = relmat), data_snpcov = gdat, format_snpcov = "-11", batch_size = 100, cores = cores)


data("assoc.FRI.rrblup", package = "athaliana") 
assoc <- left_join(as_data_frame(assoc), annot, by = "snp")

assoc_common <- assoc %>% filter(mac_FRI >= 20)

kable(head(assoc))

qq(assoc$pval)
manhattan(assoc, chr = "chr", bp = "pos", p = "pval", snp = "snp",
  suggestiveline = FALSE, genomewideline = -log10(0.05 / nrow(assoc)))

qq(assoc_common$pval)
manhattan(assoc_common, chr = "chr", bp = "pos", p = "pval", snp = "snp",
  suggestiveline = FALSE, genomewideline = -log10(0.05 / nrow(assoc_common)))
```


Also a quote using `>`:

> "He who gives up [code] safety for [code] speed deserves neither."
([via](https://twitter.com/hadleywickham/status/504368538874703872))
