---
title: "GWAS on the A. thialana data set"
author: "Andrey Ziyatdinov"
date: "`r Sys.Date()`"
output:
  rmarkdown::html_vignette:
    fig_caption: yes
vignette: >
  %\VignetteIndexEntry{GWAS on the A. thialana data set}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r options, echo = F}
opts_chunk$set(comment = NA, results = 'markup', tidy = F, message = F, warning = F, echo = T,
  fig.width = 3, fig.height = 3)
```


## Settings

### Parameters

```{r par}
cores <- 2
```

### Packages

```{r inc}
library(dplyr)
library(Matrix)

library(ggplot2)

library(rrBLUP)
library(qqman)
```

```{r inc2}
library(devtools)
load_all("~/git/variani/athaliana")
load_all("~/git/hemostat/lme4qtl")
```

### Settings

```{r settings}
theme_set(theme_light())
```

## Load the data

### Phenotypes

```{r load_phen, cache = T}
phen <- athaliana_phen(traits = "FRI", rows_order = "snp")
```

```{r tab_phen, echo = FALSE, results = "asis"}
kable(head(phen))
```

```{r transofrm_phen}
phen <- mutate(phen, 
  log_FRI = log(FRI))
```

### Pre-computed GRM

```{r data_relmat, cache = T}
relmat <- athaliana_relmat()
```

```{r plot_relmat, fig.width = 3, fig.height = 3, fig.show = "hold", cache = TRUE, dependson = "data_relmat"}
image(Matrix(relmat))
image(Matrix(relmat[1:5, 1:5]))
```

```{r tab_relmat, echo = FALSE, results = "asis"}
kable(relmat[1:5, 1:5], digits = 2)
```

### Genotypes

```{r gdata, cache = TRUE}
gdat <- athaliana_snp()
  
gdat_subset <- gdat[1:20]
```

```{r tab_gdat, echo = FALSE, results = "asis", cache = T}
kable(head(select(gdat, 1:5)))
```

### Annotation

```{r annot, cache = T}
annot <- athaliana_annot()
```

## Polygenic model

```{r poly, cache = T}
(poly <- relmatLmer(log_FRI ~ (1|id), phen, relmat = list(id = relmat)))
```

## Association study of a small subset of snps

### eigen-based model

```{r run_GWAS}
run_GWAS <- function(phen, gdat, ...) 
{
  gmat <- t(gdat[-1])
  ids <- gdat[[1]]
  colnames(gmat) <- ids
  
  pheno <- phen[c("ecotype_id", "log_FRI")]
  geno <- bind_cols(data_frame(snp = names(gdat)[-1], chr = 0, pos = 0), as_data_frame(gmat))
  
  GWAS(as.data.frame(pheno), as.data.frame(geno), ...)
}  
```

```{r assoc_mm, cache = T} 
assoc_mm <- as_data_frame(run_GWAS(phen, gdat_subset, P3D = FALSE, n.core = cores, K = relmat, plot = FALSE))
```

### lmer model

```{r assoc_lmer, cache = T}
assoc_lmer <- assocLmer(log_FRI ~ (1|id), phen, relmat = list(id = relmat), 
  data_snpcov = gdat_subset, method = "Wald",
  batch_size = 10, cores = cores)
```


## Association study on all snps

```{r load_gwas, cache = T, echo = F}
dir <- "/home/andrey/git/variani/athaliana/results/gwas/FRI"
load(file.path(dir, "gwas_rrblup_emmax.RData"))
load(file.path(dir, "gwas_rrblup_mm.RData"))
load(file.path(dir, "gwas_lmer_wald.RData"))
load(file.path(dir, "gwas_lmer_lr.RData")) 
```

```{r join_gwas, cache = T, echo = F}
tab <- data_frame(marker = gwas_rrblup_emmax$marker, 
  chrom = gwas_rrblup_emmax$chrom, pos = gwas_rrblup_emmax$pos,
  pval_emmax = 10^(-gwas_rrblup_emmax$log_FRI),
  pval_mm = 10^(-gwas_rrblup_mm$log_FRI))

tab <- bind_cols(tab, data_frame(snp = gwas_lmer_wald$snp, pval_wald = gwas_lmer_wald$pval))
tab <- bind_cols(tab, data_frame(pval_lr = gwas_lmer_lr$pval))

tab <- filter(tab, pval_emmax < 1 | pval_mm < 1)
```  


### Manhattan plot

```{r man, cache = T, fig.width = 6, fig.height = 4, echo = F}
manhattan(tab, chr = "chrom", bp = "pos", p = "pval_wald", snp = "snp", suggestiveline = FALSE, genomewideline = -log10(0.05 / nrow(tab)))
```

### QQ plot

```{r qq, fig.width = 6, fig.height = 6, cache = T, echo = F}
qq(tab$pval_wald)
```

### Are p-values consistent across methods?

#### `mm` _vs._ `emmax`

The figure below compares two methods `mm` and `emmax` in terms of diffence in p-values at log scale.
Given that `mm` is an exact method and `emmax` is an approximation,
the left panel shows that `emmax` underestimates the significance of effects.

The right panel points out that this is generally the case for the most significant effects
(located on the right to the vertical line at `-log10(pval)` equal to `5`).
These SNPs are likely to influence the estimation of variance components, i.e. heritability.

```{r plot_pval1, fig.width = 3, fig.height = 3, fig.show = "hold", cache = TRUE, echo = F}
ggplot(tab, aes(-log10(pval_emmax), -log10(pval_mm))) + geom_point() +
  geom_abline(slope = 1, intercept = 0, linetype = 3) + coord_equal()

ggplot(tab, aes(-log10(pval_mm), -log10(pval_mm/pval_emmax))) + geom_point() +
  geom_hline(yintercept = 1, linetype = 2) + geom_vline(xintercept = 5, linetype = 2)
```

#### `mm` _vs._ `wald`

Two methods `mm` and `wald` follows the same approach of using mixed models estimated for every SNP,
and the Wald test is used to derive the p-value of association.

The figure shows the perfect match of the two methods. 
That means two implementations of the method in R packages `rrBLUP` and `lme4qtl` are consistent.

```{r plot_pval2, fig.width = 3, fig.height = 3, fig.show = "hold", cache = TRUE, echo = F}
ggplot(tab, aes(-log10(pval_wald), -log10(pval_mm))) + geom_point() +
  geom_abline(slope = 1, intercept = 0, linetype = 3) + coord_equal()
```

#### `wald` _vs._ `lr`

Two mehtods implemented in the `lme4qtl` package employs different tests in the mixed model fraework:
the Wald test and the LRT.

The figure below shows the different in p-values for these two methods.

```{r plot_pval3, fig.width = 3, fig.height = 3, fig.show = "hold", cache = TRUE, echo = F}
ggplot(tab, aes(-log10(pval_lr), -log10(pval_wald))) + geom_point() +
  geom_abline(slope = 1, intercept = 0, linetype = 3) + coord_equal()

ggplot(tab, aes(-log10(pval_wald), -log10(pval_wald/pval_lr))) + geom_point() +
  geom_hline(yintercept = 1, linetype = 2) + geom_vline(xintercept = 5, linetype = 2)
```

The previous figure is not enough to get an idea about the distribution of differences.
The two histograms give a clue on what is happening.
One histogram on the left is for all the SNPs,
and another on the right is for a subset of SNPs 
where the difference in p-values at log scale is between `-2` and `2`,

```{r plot_pval3_hist, fig.width = 3, fig.height = 3, fig.show = "hold", cache = TRUE, echo = F}
ggplot(tab, aes( -log10(pval_wald/pval_lr))) + geom_histogram()

ggplot(tab, aes( -log10(pval_wald/pval_lr))) + geom_histogram() + xlim(c(-2, 2))
```

How many SNPs have a relatively large diffrenece in p-values at log scale?
`r with(tab, sum(abs(-log10(pval_wald / pval_lr)) > 1))` SNPs have the absolute value more than `1`,
and 
`r with(tab, sum(abs(-log10(pval_wald / pval_lr)) > 2))` SNPs have the absolute value more than `2`.
The total number of SNPs is `r nrow(tab)`.

