---
title: "GWAS on the A. thialana data set"
author: "Andrey Ziyatdinov"
date: "`r Sys.Date()`"
output:
  rmarkdown::html_vignette:
    fig_caption: yes
vignette: >
  %\VignetteIndexEntry{GWAS on the A. thialana data set}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

## Settings

### Parameters

```{r par}
chr <- 4

cores <- 2
```

### Packages

```{r inc}
library(dplyr)
library(Matrix)

library(athaliana)

library(lme4qtl)

library(qqman)
```

## Load the data

### Phenotypes

```{r data_phen, cache = T}
phen <- athaliana_phen(traits = "FRI", rows_order = "snp")
```

```{r tab_phen, echo = FALSE, results = "asis"}
kable(head(phen))
```

### Pre-computed GRM

```{r data_relmat, cache = T}
relmat <- athaliana_relmat()
```

```{r plot_relmat, fig.show = "hold", cache = TRUE, dependson = "data_relmat"}
image(Matrix(relmat))
image(Matrix(relmat[1:5, 1:5]))
```

```{r tab_relmat, echo = FALSE, results = "asis"}
kable(relmat[1:5, 1:5], digits = 2)
```

### Genotypes

```{r gdata, cache = TRUE}
gdat <- athaliana_snp(chr = chr)
  
gdat_subset <- gdat[1:20]
```

```{r tab_gdat, echo = FALSE, results = "asis"}
kable(head(select(gdat, 1:5)))
```

### Annotation

```{r annot, cache = T}
annot <- athaliana_annot()
```

## Polygenic model

```{r poly, cache = T}
(poly <- relmatLmer(FRI ~ (1|id), phen, relmat = list(id = relmat)))
```

## Association study of a small subset of snps

```{r assoc, cache = T}
snps <- names(gdat_subset) %>% grep("^snp", ., value = TRUE)

assoc <- assocLmer(FRI ~ (1|id), phen, relmat = list(id = relmat), 
  data_snp_cov = subset(gdat_subset, select = snps), 
  batch_size = 10, cores = cores)
```


## Association study on all snps

```{r gwas, eval = F}
snps <- names(gdat) %>% grep("^snp", ., value = TRUE)

assoc <- assocLmer(FRI ~ (1|id), phen, relmat = list(id = relmat), 
  data_snp_cov = subset(gdat, select = snps),
  batch_size = 100, cores = cores)
```

```{r load_gwas, cache = T}
data("assoc.FRI", package = "athaliana") 

assoc <- left_join(as_data_frame(assoc), annot, by = "snp")
```

```{r assoc_common, cache = T}
assoc_common <- assoc %>% filter(mac_FRI >= 20)
```

```{r tab_gwas, echo = FALSE, results = "asis", dependson = "load_gwas"}
kable(head(assoc))
```

```{r plot_gwas, fig.show = "hold", cache = T, dependson = "load_gwas"}
qq(assoc$pval)
manhattan(assoc, chr = "chr", bp = "pos", p = "pval", snp = "snp",
  suggestiveline = FALSE, genomewideline = -log10(0.05 / nrow(assoc)))
```

```{r plot_gwas_common, fig.show = "hold", cache = T, dependson = "load_gwas"}
qq(assoc_common$pval)
manhattan(assoc_common, chr = "chr", bp = "pos", p = "pval", snp = "snp",
  suggestiveline = FALSE, genomewideline = -log10(0.05 / nrow(assoc_common)))
```

Also a quote using `>`:

> "He who gives up [code] safety for [code] speed deserves neither."
([via](https://twitter.com/hadleywickham/status/504368538874703872))
